lib/OpenQA/WebAPI/Controller/API/V1/Iso.pm

```diff
@@ -15,13 +15,390 @@
+                        });
+                }
+            }
+        }
+    }
+}
+
+
+# internal function not exported - but called by create
+sub schedule_iso {
+    my ($self, $args) = @_;
+    # register assets posted here right away, in case no job
+    # templates produce jobs.
+    for my $a (values %{parse_assets_from_settings($args)}) {
+        OpenQA::Scheduler::Scheduler::asset_register(%$a);
+    }
+    my $noobsolete = delete $args->{_NOOBSOLETEBUILD};
+    # Any arg name ending in _URL is special: it tells us to download
+    # the file at that URL before running the job
+    my %downloads = ();
+    for my $arg (keys %$args) {
+        next unless ($arg =~ /_URL$/);
+        # As this comes in from an API call, URL will be URI-encoded
+        # This obviously creates a vuln if untrusted users can POST
+        $args->{$arg} = url_unescape($args->{$arg});
+        my $url = $args->{$arg};
+        # if $args{FOO_URL} is set but $args{FOO} is not, we will
+        # set $args{FOO} (the filename of the downloaded asset) to
+        # the URL filename. This has to happen *before*
+        # generate_jobs so the jobs have FOO set
+        my $short = substr($arg, 0, -4);
+        my $assettype = asset_type_from_setting($short);
+        # We're only going to allow downloading of asset types
+        unless ($assettype) {
+            OpenQA::Utils::log_debug("_URL downloading only allowed for asset types! $short is not an asset type");
+            next;
+        }
+        if (!$args->{$short}) {
+            $args->{$short} = Mojo::URL->new($url)->path->parts->[-1];
+            if (!$args->{$short}) {
+                OpenQA::Utils::log_warning("Unable to get filename from $url. Ignoring $arg");
+                delete $args->{$short} unless $args->{$short};
+                next;
+            }
+        }
+        # full path to download target location. We need to guess
+        # the asset type to know where to put it, using the same
+        # subroutine as parse_assets_from_settings
+        my $dir = catdir($OpenQA::Utils::assetdir, $assettype);
+        my $fullpath = catfile($dir, $args->{$short});
+
+        unless (-s $fullpath) {
+            # if the file doesn't exist, add the url/target path
+            # as a key/value pair to the %downloads hash
+            $downloads{$url} = $fullpath;
+        }
+    }
+    my $jobs = $self->_generate_jobs($args);
+
+    # XXX: take some attributes from the first job to guess what old jobs to
+    # cancel. We should have distri object that decides which attributes are
+    # relevant here.
+    if (!$noobsolete && $jobs && $jobs->[0] && $jobs->[0]->{BUILD}) {
+        my %cond;
+        for my $k (qw/DISTRI VERSION FLAVOR ARCH/) {
+            next unless $jobs->[0]->{$k};
+            $cond{$k} = $jobs->[0]->{$k};
+        }
+        if (%cond) {
+            OpenQA::Scheduler::Scheduler::job_cancel(\%cond, 1);    # have new build jobs instead
```

Comment:
I actually wonder if we can do the job cancelling as a side job in the scheduler / websocket handler. As it looks we're waiting for the old jobs to cancel - but it's far from important.